crates/apub/src/http/inbox.rs

```diff
@@ -0,0 +1,171 @@
+use lemmy_utils::error::LemmyResult;
+use once_cell::sync::Lazy;
+use std::{
+  cmp::Ordering,
+  collections::BinaryHeap,
+  sync::{Arc, RwLock},
+  thread::available_parallelism,
+  time::Duration,
+};
+use tokio::{spawn, task::JoinHandle, time::sleep};
+use tracing::info;
+
+/// Handle incoming activities.
+pub async fn shared_inbox(
+  request: HttpRequest,
+  bytes: Bytes,
+  data: Data<LemmyContext>,
+) -> LemmyResult<HttpResponse> {
+  match serde_json::from_slice::<WithPublished<()>>(&bytes)?.published {
+    Some(published) => {
+      // includes published timestamp, insert to queue to ensure that activities are processed
+      // in correct order even when delivered out of order.
+      let request_parts = (
+        request.headers().clone(),
+        request.method().clone(),
+        request.uri().clone(),
+      );
+      ACTIVITY_QUEUE.write().unwrap().push(InboxActivity {
+        request_parts,
+        bytes,
+        published,
+      });
+      Ok(HttpResponse::Ok().finish())
+    }
+    None => {
+      // no timestamp included, process immediately
+      receive_activity::<SharedInboxActivities, UserOrCommunity, LemmyContext>(
+        request, bytes, &data,
+      )
+      .await
+    }
+  }
+}
+
+/// Queue of incoming activities, ordered by oldest published first
+static ACTIVITY_QUEUE: Lazy<Arc<RwLock<BinaryHeap<InboxActivity>>>> =
+  Lazy::new(|| Arc::new(RwLock::new(BinaryHeap::new())));
+
+/// Minimum age of an activity before it gets processed. This ensures that an activity which was
+/// delayed still gets processed in correct order.
+const RECEIVE_DELAY: Option<TimeDelta> = TimeDelta::try_seconds(1);
+
+pub fn handle_received_activities(
+  context: &Data<LemmyContext>,
+) -> LemmyResult<Vec<JoinHandle<()>>> {
+  // launch one task per cpu core
+  let parallelism = available_parallelism()?.into();
+  let workers = (0..parallelism)
+    .map(|_| {
+      let context = context.reset_request_count();
+      spawn(async move {
+        loop {
+          let now = Local::now();
+          if let Some(latest_timestamp) = peek_queue_timestamp() {
+            if latest_timestamp < now - RECEIVE_DELAY.unwrap() {
+              if let Some(a) = pop_queue() {
+                let parts = (&a.request_parts.0, &a.request_parts.1, &a.request_parts.2);
+                receive_activity_parts::<SharedInboxActivities, UserOrCommunity, LemmyContext>(
+                  parts, a.bytes, &context,
+                )
+                .await
+                .inspect_err(|e| info!("Error receiving activity: {e}"))
+                .ok();
+              }
+            }
+          }
+          // TODO: could sleep based on remaining time until head activity reaches 1s
+          //       or simply use `WORK_FINISHED_RECHECK_DELAY` from lemmy_federate
+          sleep(Duration::from_millis(100)).await;
+          // TODO: need cancel? lemmy seems to shutdown just fine
+        }
+      })
+    })
+    .collect();
```

Comment:
mhh. This is very similar to what I imagined, but it's different in a major way that might need consideration.

The main thing is that you only have a single receive queue but within that queue do fully parallel processing of activities (with delay). I had imagined one fully sequential queue per remote instance, with each queue being processed in parallel.

A single queue has the problem that you still don't guarantee the dependencies are processed in order. Even with the delay, let's say you have these events in the queue:

- T-1.01s: user N upvote post X
- T-1.00s: user N downvote post X

The activities are both received correctly in the queue, and regardless of receive order have been priority-ordered by the timestamp. . But the problem is that both can be taken by different of your `available_parallelism` workers at the same time (The 1s delay has been met so both are up for processing). That means the downvote might be processed before the upvote and then wrongly overwritten.

With my suggestion both would be processed sequentially which means this problem doesn't happen.

The advantage of your method is that it also solves a slow rust to database connection (like [db0](https://dbzer0.com/blog/post-mortem-the-massive-lemmy-world-lemmy-dbzer0-com-federation-delays/) had) because the internal latency doesn't matter. But the delay alone doesn't really solve the sequentiality problem, so the result can be wrong.

Sequential per-instances queues still solves #4529 the roundtrip between the instances is not part of the latency, only the internal latency remains. 

I'm not sure what the best way forward is though. I do see now looking at this code that per-instance receive queues would be fairly complex.

Reply:
I have a different concern. Consider a cluster of X lemmy servers; this _cannot_ be handled in process if order matters and lemmy is intended to scale beyond a single process.

This is something that would normally be handled by an out of process job queue like kafka, sidekiq, rabbitmq, etc.

You push the job into the message queue, response 200, and then move on, on the web server side. Then a (potential) cluster of workers pops off the job from the queue in the background.

Those workers can then either post back to the sender (in the case of lemmy the sending instance) "I've done the work you wanted" or the sender can be given a ticket to check back in with periodically (if there's truly an order that needs enforced on the sending side).

I think that latter strategy (where the sender paces itself based on status updates from the receiver) is out based on what I'm reading as we're working within the constraints of ActivityPub which means your queue needs to be smart enough to block where appropriate and not take the wrong jobs.

I think at a minimum this would need to use PostgreSQL (without adding software to the lemmy stack) to implement a job queue table that various lemmy servers could pull from when they're not otherwise busy to handle background work. That could also be used to resolve your (@phiresky) concern about multiple jobs being taken at the same time by tracking dependence, and not flagging the latter job for execution until the earlier job is done.

Reply:
I'm not sure how practical this is but, it seems like the ideal solution would be to make this eventually consistent by using the timestamp not for ordering but to make accept/reject judgements.

```
T-1.01s: user N upvote post X
T-1.00s: user N downvote post X
```

In that case, you receive the first event, it's the most recent state for `user N` on `post X` so you take that as new current state for `user N`. Then you see the second event which happened in the past, and upon processing disregard it as irrelevant.

That can also open the door to a bunch of different potential race condition situations between complex endpoints and database state though (where e.g., two threads are working on changes affecting the same posts/likes/etc at the same time). So, that would _probably_ have to be done selectively based on the "type"(?) of activity.

It might be helpful if such a thing could be done for e.g., likes as they're the most likely to be high frequency requests I'd imagine.

Reply:
Race conditions shouldn't be a problem if old events are ignored by doing the upsert with a `WHERE` clause that compares the existing row's timestamp. `UPDATE` will redo the comparison if needed after locking the row.

https://www.postgresql.org/docs/current/transaction-iso.html

Reply:
> Race conditions shouldn't be a problem if old events are ignored by doing the upsert with a `WHERE` clause that compares the existing row's timestamp. `UPDATE` will redo the comparison if needed after locking the row.
> 
> https://www.postgresql.org/docs/current/transaction-iso.html

This is true for anything that's represented via a single row. The concern would be more about anything that triggers a complex update that involves multiple tables and/or rows. Perhaps that isn't much of a concern currently though.

Reply:
I just realized that using timestamps to reject older events will require some tables to be changed since removal timestamps are not stored for many things. This should be done after #4459 is merged so there's no conflict (or maybe in the same PR).

Reply:
Your comment a out multiple chances is a good point why an in memory queue is probably a bad idea. I think it's possible to solve this whole thing only on the outgoing side without requiring any further changes with what I've mentioned in point 9 of https://github.com/LemmyNet/lemmy/issues/4529#issuecomment-2015272947

Reply:
If we process incoming activities sequentially then we will again run into the same problem sooner or later. Then not caused by high ping but because one instance is sending more activities than the other one can handle in the timeframe. So parallel processing is definitely necessary, also to utilize all available server resources when catching up after an instance fell back.

Reply:
for reference, what @DarkArc described above is I think the same thing as what I called "(5) making all activities commutative" in my list in #4529